dagshub/streaming/dataclasses.py

```diff
@@ -69,57 +60,37 @@ def __post_init__(self):
         if self.relative_path is not None:
             str_path = self.relative_path.as_posix()
             for storage_schema in storage_schemas:
-                if str_path.startswith(storage_schema):
-                    str_path = str_path[len(f"{storage_schema}"):]
-                    self.relative_path = Path(".dagshub/storage") / storage_schema[:-2] / str_path
+                if str_path.startswith(f"{storage_schema}:/"):
+                    str_path = str_path[len(storage_schema) + 2:]
```

Comment:
I still don't understand how that takes care of `fs.open("s3://bucket/path/to/file")`.

Reply:
DagshubPath is a path abstraction that carries in it:
- Absolute path in the FS
- Path relative to the repo root (if it's in the repo)

**1) Why don't I handle it as URIs**
Because when this object is created, it's getting Paths, so the original requested string has already went through some "mangling".
You can open a Python interpreter and see it for yourself
```
>>> from pathlib import Path
>>> a = Path("s3://hahaha")
>>> a.as_posix()
's3:/hahaha'
```

So by the time this gets to DagshubPath this is an invalid URI already.
Yes, I can instead pass strings into DagshubPath, but that'd require me to transfer all of the relative pathing logic into here also, which I'm not sure if I want to. I want to keep as little logic here as I could.
However...

**2) HOW is it handling the URI**
There _is_ a bit of logic here
The code you commented on is actually a `__post_init__()` function, which is called in dataclasses right after initialization.
So what is going here is after that object is initialized, I check if `relative_path` starts with a storage schema, and if it does, then I force-translate it to actually be at the path I need it to be at (and also change the absolute path along the way).
This way, right after DagshubPath has been initialized with "s3://test/file", any function that later handles it will actually get a DagshubPath that points to `.dagshub/storage/s3/test/file`

Reply:
> ```python
>>> from pathlib import Path
>>> a = Path("s3://hahaha")
>>> a.as_posix()
> 's3:/hahaha'
```

This is the part I was missing. Thanks for clarifying.